{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "effd9467-fe12-4cbc-96eb-a1b84789147c",
   "metadata": {},
   "source": [
    "# Lecture 7 - Functional gene annotation\n",
    "\n",
    "In this lecture you learned about annotating gene function using controlled vocabularies like **EC numbers** and **GO terms**.\n",
    "\n",
    "### Learning objectives:\n",
    "\n",
    "- Learn to functionally annotate genes with InterProScan\n",
    "- Become familiar with the Pandas library\n",
    "- Navigating ontologies with recursive functions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d6219deb-ffa2-46ab-b2d8-fcc9261e972f",
   "metadata": {},
   "source": [
    "## Exercise 1\n",
    "\n",
    "In lecture 2 we assembled the genome of [*Mycoplasma pneumoniae*](https://en.wikipedia.org/wiki/Mycoplasma_pneumoniae) from sequencing data downloaded from the European Nucleotide Archive (accession [DRR040043](https://www.ebi.ac.uk/ena/browser/view/DRR040043?show=reads)).\n",
    "\n",
    "In lecture 3 we used **Prodigal** to find all open reading frames in the assembled genome and translate them to amino acid sequences.\n",
    "\n",
    "In this exercise, we will functionally annotate those amino acid sequences using [InterProScan](https://www.ebi.ac.uk/interpro/about/interproscan/). This tool scans a query protein sequence against all protein families in InterPro (using HMM profiles as discussed in lecture 5).\n",
    "\n",
    "The search engine for **InterProScan** works similarly to **BLAST**, using the [web portal](https://www.ebi.ac.uk/interpro/search/sequence/) you can only submit up to 100 sequences at once. Fortunately, there is an instance of **InterProScan** available in **Galaxy** that allows to annotate all the sequences in a FASTA file:\n",
    "\n",
    "![galaxy](files/interproscan.png)\n",
    "\n",
    "* Upload the [proteins.faa](files/proteins.faa) file (you must first download it from *lecture_07/files*)\n",
    "* Search for *InterProScan* and select the file you just uploaded\n",
    "* Make sure all databases are selected\n",
    "* Select **Run Tool**\n",
    "\n",
    "ðŸ‘‰ This will take about 30 minutes to run. In the next exercise we will use a pre-computed result. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0660b45b-8fa8-487a-b0d9-f5f32cbcab36",
   "metadata": {},
   "source": [
    "## Exercise 2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d17d37b-1609-4f1b-b166-e8706a264522",
   "metadata": {},
   "source": [
    "The annotation result is a TSV (tab-separated values) file that you can very easily load with **Pandas**.\n",
    "\n",
    "ðŸ‘‰ **Pandas** is an extremely powerful library for working with tabular data (including Excel files). I strongly encourage you to spend some minutes looking at the [documentation](https://pandas.pydata.org/docs/index.html).\n",
    "\n",
    "Let's start by loading the results:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0fe31728-5de0-401e-8298-2be0fdd24362",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "df = pd.read_csv('files/results.tsv', sep='\\t', header=None)\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c24fb9e9-0e66-4e55-a4dc-e6906a74ddef",
   "metadata": {},
   "source": [
    "As you can see there are several columns. According to the [InterProScan documentation](https://interproscan-docs.readthedocs.io/en/v5/OutputFormats.html#tab-separated-values-format-tsv) the output format is as follows:\n",
    "\n",
    "1. Protein accession \n",
    "2. Sequence MD5 digest \n",
    "3. Sequence length \n",
    "4. Analysis (e.g. Pfam / PRINTS / Gene3D)\n",
    "5. Signature accession \n",
    "6. Signature description \n",
    "7. Start location\n",
    "8. Stop location\n",
    "9. Score - is the e-value (or score) \n",
    "10. Status - is the status of the match \n",
    "11. Date - is the date of the run\n",
    "12. InterPro annotations - accession\n",
    "13. InterPro annotations - description\n",
    "14. GO annotations with their source(s)\n",
    "15. Pathways annotations \n",
    "\n",
    "ðŸ‘‰ Note that the columns in Pandas are numbered 0 to 14.\n",
    "\n",
    "Let's only keep the query protein, target database, and annotated GO terms:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5350ec39-7114-453f-8bef-c8a00fb61121",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv('files/results.tsv', sep='\\t', header=None, usecols=[0, 3, 13], na_values='-', names=['query', 'database', 'GO']).dropna()\n",
    "df.sample(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c6097e2-a840-4928-ad21-6abe31676e84",
   "metadata": {},
   "source": [
    "Let's also *\"unpack\"* the GO terms, so that we only have one term per row (this will replicate the values in the other columns):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8cbcc54-5c1f-475d-8ca9-a5c9d08491bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['GO'] = df['GO'].str.split('|')\n",
    "df = df.explode('GO').drop_duplicates()\n",
    "df.sample(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "37f0662e-b3b7-4f2f-be01-760aad67824e",
   "metadata": {},
   "source": [
    "We can use `.value_counts()` to find the most frequent values in each column:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "07cc3c9a-2dd3-4ae1-92af-05926df96bad",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['database'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a87bd66f-9ad7-4b7e-a0a5-77fe14c79ed2",
   "metadata": {},
   "source": [
    "### 2.1\n",
    "\n",
    "What is the average (median) number of GO terms annotated for each protein?\n",
    "\n",
    "> ðŸ’¡ Tip: you can combine the results of multiple operations over pandas dataframes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1bc33da-88c0-4167-9cd4-d14e77925966",
   "metadata": {},
   "outputs": [],
   "source": [
    "# type your code here"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7a13346c-0832-4e84-ba08-f4be0254f505",
   "metadata": {},
   "source": [
    "Click below to see a solution..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e4888f2-a161-40e2-ba0c-135622f80a38",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "x = df['query'].value_counts().median()\n",
    "print(f'The answer is: {x:n}') # the :n here is just formatting the float as an integer 'n'umber"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "226576cf-d401-4582-9692-4840b4fee26c",
   "metadata": {},
   "source": [
    "### 2.2\n",
    "\n",
    "Can you print the top 10 most frequent GO terms?\n",
    "\n",
    "> ðŸ’¡ Tip: the `.head()` method can be useful."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aad7a004-9bcb-4626-9b48-0da0f804d548",
   "metadata": {},
   "outputs": [],
   "source": [
    "# type your code here..."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a9a3d91a-c6b9-476e-be14-2da705f1e09e",
   "metadata": {},
   "source": [
    "Click below to see a solution..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aba5b09b-ff8e-4aeb-9433-2ac8fc090e4e",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "df['GO'].value_counts().head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b9e3aa1-3e94-4953-8a64-ccd62b0c01cb",
   "metadata": {},
   "source": [
    "---------\n",
    "\n",
    "## Exercise 3 - Gene Ontology\n",
    "\n",
    "[**GOATOOLS**](https://www.nature.com/articles/s41598-018-28948-z) is a useful Python library to work with the [Gene Ontology](https://geneontology.org/).\n",
    "\n",
    "Let's start by loading the latest version (well... at least it was when I wrote this) of the complete ontology from a local file (downloaded from [here](https://geneontology.org/docs/download-ontology/)):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7797a431-5766-44d5-bc70-6e4d1a36b063",
   "metadata": {},
   "outputs": [],
   "source": [
    "from goatools import obo_parser\n",
    "go_terms = obo_parser.GODag('files/go-basic.obo', optional_attrs='xref')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08735297",
   "metadata": {},
   "source": [
    "We can use this library to get more information about the GO terms annotated in our genome. \n",
    "\n",
    "`go_terms` is a dictionary from GO term *\"ids\"* to GO term *\"objects\"*. Let's inspect the first term..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d089e9c-89b3-46ce-816d-91850bfd2e39",
   "metadata": {},
   "outputs": [],
   "source": [
    "go_terms['GO:0000001']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a7dd490-5bdd-4bb4-9291-2a97b2a3614d",
   "metadata": {},
   "source": [
    "### 3.1\n",
    "\n",
    "InterProScan annotates with GO terms but not with EC numbers.\n",
    "\n",
    "Create a function called `go2ec()` that receives a GO term identifier and returns the respective EC number (only if that GO term represents an enzymatic reaction).\n",
    "\n",
    "> ðŸ’¡ **Tip**: You can find the respective EC number (if it exists) in the `xref` attribute and it starts with `EC:`.\n",
    "\n",
    "ðŸ‘‰ Note that you might not be able to find some GO terms in the dictionary because they are obsolete. Example: [GO:0045261](https://www.ebi.ac.uk/QuickGO/term/GO:0045261) (just ignore those)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "579bfcf8-c0a3-45e9-b803-fc1a08408301",
   "metadata": {},
   "outputs": [],
   "source": [
    "# type your code here..."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0332b040-f28c-4d09-982a-87bbe7abc244",
   "metadata": {},
   "source": [
    "Click below to see the solution:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c4ca60d0-46ec-443b-bc1b-f9808acb4f26",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "\n",
    "def go2ec(go_id):\n",
    "    if go_id in go_terms:\n",
    "        go_term = go_terms[go_id]\n",
    "        for data in go_term.xref:\n",
    "            if data.startswith('EC:'):\n",
    "                return data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe1823f9-ce06-478c-ab1b-0e18513cc540",
   "metadata": {},
   "source": [
    "Let's test your solution:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b3fbba57-4fc1-4b6c-a43f-656e7191cf30",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['EC'] = df['GO'].apply(go2ec)\n",
    "df.dropna().sample(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be67d646-eb3d-4cc6-b1c2-e3ce6247114c",
   "metadata": {},
   "source": [
    "### 3.2\n",
    "\n",
    "Remember that GO terms are organized in a hierarchy. Each GO term contains the attributes `.parents` and `.children` with a *set* of GO terms that are (immediately) above or below. \n",
    "\n",
    "Create a loop that prints a list (including id and name) of all the ancestors (parents, grandparents, ...) of the most frequent GO term that you found in exercise 2.2.\n",
    "\n",
    "> ðŸ’¡ **Tip**: this looks like a job for a recursive function (*i.e.* a function that invokes itself)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "afc5c863-26e3-4b77-9373-98db4721e048",
   "metadata": {},
   "outputs": [],
   "source": [
    "# type your code here..."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b51f445-6b0f-42e8-958f-40a57bf0aff7",
   "metadata": {},
   "source": [
    "Click below to see the solution:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ee83335-1d29-4e8b-805a-871c870ff2f5",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "\n",
    "def get_parents(go_term, indent=0):\n",
    "    # indent is just a visual aid, you can ignore that\n",
    "    spacing = ' '*14*indent\n",
    "    \n",
    "    for parent in go_term.parents:\n",
    "        print(f\"{spacing}{go_term.id} -> {parent.id}: {parent.name}\")\n",
    "        get_parents(parent, indent+1)\n",
    "\n",
    "go = go_terms['GO:0005524']\n",
    "get_parents(go)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f6d9bab-66d8-4686-acf1-87ea40100468",
   "metadata": {},
   "source": [
    "### 3.3\n",
    "\n",
    "Inspect the tree of your term in [QuickGO](https://www.ebi.ac.uk/QuickGO/) to confirm that you got all the correct terms. \n",
    "\n",
    "- ðŸ¤” Did you get a similar tree ?  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ea2e02e-5dff-4f95-b0d0-ca79777ca1b2",
   "metadata": {},
   "source": [
    "## Wrap-up\n",
    "\n",
    "This session had a lot of new things going on. If you got stuck in an exercise, just ask for help or, if you found the exercises too simple, maybe *you* can help someone. ðŸ˜‰"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.9"
  },
  "vscode": {
   "interpreter": {
    "hash": "9f894c7364278704108a83365f2f09f6b409d9aa4f470696e61cf208042e86d6"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
