{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "14c50ca2-0991-4aff-87af-5ae3ec6d232a",
   "metadata": {},
   "source": [
    "# Lecture 7 - Functional gene annotation\n",
    "\n",
    "In this lecture you have learned about finding and describing the function of genes and proteins.\n",
    "\n",
    "In lecture 3, we used [Prokka](https://academic.oup.com/bioinformatics/article/30/14/2068/2390517) to identify all the genes (open reading frames) present in an assembled genome. In addition to identifying gene sequences, Prokka also annotates those sequences, using homology-based annotation transfer, by BLASTing them against reference databases (UniProt, RefSeq, Pfam).\n",
    "\n",
    "Let's load the annotations file (in *Genbank* format) generated by Prokka. This file format is more detailed than a simple FASTA file, and contains several annotated **features** for each gene sequence:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0fe31728-5de0-401e-8298-2be0fdd24362",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "sequences = list(SeqIO.parse('../lecture_03/files/output/annotated.gbk', 'genbank'))\n",
    "\n",
    "print(sequences[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d45a1077-f3fa-4d35-b559-58740cc1e06a",
   "metadata": {},
   "source": [
    "-------\n",
    "\n",
    "## Exercise 1:\n",
    "\n",
    "Among the features that Prokka can annotate are EC numbers (unfortunately it does not annotate GO terms). \n",
    "\n",
    "Create a list (or a set, if you want to avoid repetitions) of all EC numbers annotated in this genome assembly.\n",
    "\n",
    "> How many (unique) EC numbers did you find ?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce74fba1-b978-4f40-997c-93f3e9e46bb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# type your code here..."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "66fb3fc6-c907-4ff1-9783-0da0f4cd0723",
   "metadata": {},
   "source": [
    "Click the cell below to see the solution:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d05422f-ddc7-4847-a081-07ed85dac5ae",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "ec_numbers = set()\n",
    "\n",
    "for seq in sequences:\n",
    "    for feature in seq.features:\n",
    "        if 'EC_number' in feature.qualifiers:\n",
    "            ec_numbers.update(feature.qualifiers['EC_number'])\n",
    "            \n",
    "print(len(ec_numbers))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b9e3aa1-3e94-4953-8a64-ccd62b0c01cb",
   "metadata": {},
   "source": [
    "---------\n",
    "\n",
    "## Gene Ontology\n",
    "\n",
    "[GOATOLS](https://www.nature.com/articles/s41598-018-28948-z) is a useful Python library to work with the Gene Ontology.\n",
    "\n",
    "Let's begin by installing it..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3251fe64-893f-45ae-b6da-2bcfed9f7a96",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "!pip install goatools"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "34c0bc9e-1c4d-4cb1-bf88-c5e137308f23",
   "metadata": {},
   "source": [
    "Now let's load the latest (well, it was when I wrote this) version of the complete ontology from a local file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7797a431-5766-44d5-bc70-6e4d1a36b063",
   "metadata": {},
   "outputs": [],
   "source": [
    "from goatools import obo_parser\n",
    "go_terms = obo_parser.GODag('files/go-basic.obo', optional_attrs='xref')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d089e9c-89b3-46ce-816d-91850bfd2e39",
   "metadata": {},
   "outputs": [],
   "source": [
    "go_terms['GO:0000001']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a7dd490-5bdd-4bb4-9291-2a97b2a3614d",
   "metadata": {},
   "source": [
    "## Exercise 2.1:\n",
    "\n",
    "You might notice that among the list of EC numbers detected in the genome, was our good old friend **2.7.1.11** (*PFK1*, obviously). Can you find the respective GO term?\n",
    "\n",
    "> Tips: `go_terms` is a dictionary from GO term ids to GO term objects, you can find the respective EC number (if it exists) in the `xref` attribute and it starts with `EC:`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "579bfcf8-c0a3-45e9-b803-fc1a08408301",
   "metadata": {},
   "outputs": [],
   "source": [
    "# type your code here..."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0332b040-f28c-4d09-982a-87bbe7abc244",
   "metadata": {},
   "source": [
    "Click below to see the solution:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c4ca60d0-46ec-443b-bc1b-f9808acb4f26",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "for go_id, go_term in go_terms.items():\n",
    "    if 'EC:2.7.1.11' in go_term.xref:\n",
    "        print(go_term)\n",
    "        break"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be67d646-eb3d-4cc6-b1c2-e3ce6247114c",
   "metadata": {},
   "source": [
    "## Exercise 2.2:\n",
    "\n",
    "Each GO term contains an attribute `.parents` with a list of its (immediate) parents. Can you create a loop that prints a list (including id and name) of all the ancestors? \n",
    "\n",
    "> Tip: this looks like a job for a recursive function (*i.e.* a function that calls itself)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "afc5c863-26e3-4b77-9373-98db4721e048",
   "metadata": {},
   "outputs": [],
   "source": [
    "# type your code here..."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b51f445-6b0f-42e8-958f-40a57bf0aff7",
   "metadata": {},
   "source": [
    "Click below to see the solution:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ee83335-1d29-4e8b-805a-871c870ff2f5",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "def get_parents(go_term):\n",
    "        \n",
    "    for parent in go_term.parents:\n",
    "        print(f\"{go_term.id} -> {parent.id}: {parent.name}\")\n",
    "        get_parents(parent)\n",
    "\n",
    "go = go_terms['GO:0003872']\n",
    "get_parents(go)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f6d9bab-66d8-4686-acf1-87ea40100468",
   "metadata": {},
   "source": [
    "Inspect the tree of your term in [QuickGO](https://www.ebi.ac.uk/QuickGO/) to confirm that you got all the correct terms. \n",
    "- Did you get a similar tree? \n",
    "- What are the differences ?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5265a06-ced4-466f-8184-e8945218059a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
